src/Microsoft.TemplateEngine.Core/Conditional.cs

```diff
@@ -206,85 +290,164 @@ public int HandleMatch(IProcessorState processor, int bufferLength, ref int curr
                 {
-                    //If the elseif branch is taken, return control for replacements to be done as usual
-                    if (!_current.Evaluate(processor, ref bufferLength, ref currentBufferPosition))
+                    // 8-19 attempt to make the same as if() handling
+                    //
+                    if (_current.Evaluate(processor, ref bufferLength, ref currentBufferPosition))
                     {
                         if (_definition.WholeLine)
                         {
                             processor.SeekForwardThrough(processor.EncodingConfig.LineEndings, ref bufferLength, ref currentBufferPosition);
                         }
 
-                        if (SeekToTerminator(processor, ref bufferLength, ref currentBufferPosition, out token))
+                        if (IsTokenIndexOfType(token, ElseIfTokenActionableBaseIndex))
                         {
-                            goto BEGIN;
+                            // the elseif branch is taken. 
+                            _current.ToggleActionableOperations(true, processor);
                         }
+
+                        return 0;
                     }
+                    else
+                    {
+                        SeekToNextTokenAtSameLevel(processor, ref bufferLength, ref currentBufferPosition, out token);
 
-                    if (_definition.WholeLine)
+                        // In the original version this was conditional on SeekToToken() succeeding.
+                        // Not sure if it should be conditional. It should never fail, unless the template is malformed.
+                        goto BEGIN;
+                    }
+                }
+
+                //We have an "else" token and haven't taken any other branches, return control
+                //  after setting that a branch has been taken
+                if (IsTokenIndexOfType(token, ElseTokenBaseIndex) || IsTokenIndexOfType(token, ElseTokenActionableBaseIndex))
+                {
+                    if (IsTokenIndexOfType(token, ElseTokenActionableBaseIndex))
                     {
-                        processor.SeekForwardThrough(processor.EncodingConfig.LineEndings, ref bufferLength, ref currentBufferPosition);
+                        _current.ToggleActionableOperations(true, processor);
                     }
 
-                    //The "elseif" branch was not taken. Skip to the following else, elseif or endif token
+                    _current.BranchTaken = true;
+                    processor.WhitespaceHandler(ref bufferLength, ref currentBufferPosition, wholeLine: _definition._wholeLine, trim: _definition._trimWhitespace);
                     return 0;
                 }
+                else
+                {
+                    throw new InvalidDataException("Unknown token index: " + token);
```

Comment:
Sure, I’ll need to learn our standards for exception handling.

From: Mike Lorbetske [mailto:notifications@github.com]
Sent: Tuesday, August 23, 2016 11:12 AM
To: seancpeters/microsoft-templateengine <microsoft-templateengine@noreply.github.com>
Cc: Sean Peters <sepeters@microsoft.com>; Author <author@noreply.github.com>
Subject: Re: [seancpeters/microsoft-templateengine] Refactor (#1)


In src/Microsoft.TemplateEngine.Core/Conditional.cs<https://github.com/seancpeters/microsoft-templateengine/pull/1#discussion_r75921028>:

>                      return 0;

>                  }

> +                else

> +                {

> +                    throw new InvalidDataException("Unknown token index: " + token);

Debug assert maybe?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/seancpeters/microsoft-templateengine/pull/1/files/a425b2f9e55c8aa446463bd90060cb6bc99db3d0#r75921028>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AT8a8VEVQeKZiZguJ-wyehOAFXc7NL2Jks5qizfigaJpZM4JqaeQ>.